03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd

```diff
@@ -0,0 +1,414 @@
+Note that rows represent gene data and the columns represent sample data in our dataset.
+
+```{r}
+# Define a minimum counts cutoff and filter `DESeqDataSet` object to include
+# only rows that have counts above the cutoff
+genes_to_keep <- rowSums(counts(dds)) >= 10
+dds <- dds[genes_to_keep, ]
+```
+
+## Perform DESeq2 normalization
+
+We are going to use the `vst()` function from the `DESeq2` package to normalize the data.
+TODO: Add a note/link to section in documentation on why we use `vst()` vs `rlog()` and `normTransform()`.
+
+```{r}
+# Normalize the data in the `DESeqDataSet` object using the `vst()` function from the `DESEq2` R package
+dds_norm <- vst(dds)
+```
+
+## Perform UMAP
+
+Uniform Manifold Approximation and Projection (UMAP) is a dimension reduction technique proposed by McInnes, Healy, & Melville as it isdescribed in the [associated paper](https://arxiv.org/abs/1802.03426) [@umap-paper].
+While PCA assumes that the variation we care about has a particular distribution (normal, broadly speaking), UMAP allows more complicated distributions that it learns from the data. 
+The advantage of this feature is that UMAP can do a better job separating clusters, especially when some of those clusters may be more similar to each other than others.  
+
+In this code chunk, we are going to extract the normalized counts data from the `DESeqDataSet` object and perform UMAP on the normalized data using `umap()` from the `umap` package.
+We are using the default parameters when we run the `umap::umap()` function.
+Take a look at the `Tuning UMAP` section at [this link](https://cran.r-project.org/web/packages/umap/vignettes/umap.html) for more on parameters that can be configured when executing `umap::umap()` [@umap-parameters].
+You can also run the following in the RStudio console to get more information on the function and its default parameters: `?umap::umap` or `?umap::umap.defaults`.
+
+```{r}
+# First we are going to retrieve the normalized data from the `DESeqDataSet` object using the `assay()` function
+normalized_counts <- assay(dds_norm) %>%
+  t() # We need to transpose this data in preparation for the `umap()` function
+
+# Now let's tell R to perform UMAP on the normalized data
+umap_results <- umap::umap(normalized_counts)
+```
+
+## Prepare data frame for plotting
+
+Now that we have the results from UMAP, we need to extract the counts data from the `umap_results` object and merge the variables from the metadata that we will use for annotating our plot.
+
+```{r}
+# We need to tell R to create a data frame with the umap counts and annotation data in preparation for plotting with `ggplot2`
+umap_plot_df <- data.frame(umap_results$layout) %>%
+  tibble::rownames_to_column("refinebio_accession_code") %>% # Let's get the rownames into a column so we can merge the annotation data
+  dplyr::inner_join(metadata, by = "refinebio_accession_code") # Let's merge the annotation data using the `refinebio_accession_code` column
```

Comment:
I think there is a way to join using rownames, but honestly I find this strategy you used here easier to follow/explain.